geowave/src/main/scala/geotrellis/spark/io/geowave/GeowaveLayerReader.scala

```diff
@@ -0,0 +1,256 @@
+      LayoutDefinition(extent, tileLayout)
+    }
+
+    val cellType = metadata.get("cellType") match {
+      case null => {
+        val geom = (new GeometryFactory).createPoint(new Coordinate((minX + maxX) / 2.0, (minY + maxY) / 2.0))
+        val queryOptions = new QueryOptions(adapter, index)
+        val query = new IndexOnlySpatialQuery(geom)
+        val config = computeConfiguration
+        GeoWaveInputFormat.setQuery(config, query)
+        GeoWaveInputFormat.setQueryOptions(config, queryOptions)
+
+        val gc = sc.newAPIHadoopRDD(
+          config,
+          classOf[GeoWaveInputFormat[GridCoverage2D]],
+          classOf[GeoWaveInputKey],
+          classOf[GridCoverage2D])
+          .map({ case (_, gc) => gc })
+          .collect.head
+
+        GridCoverage2DConverters.getCellType(gc)
+      }
+      case s: String => CellType.fromString(s)
+    }
+
+    val bounds = KeyBounds(
+      SpatialKey(0, 0),
+      SpatialKey(maxCol - minCol, maxRow - minRow)
+    )
+
+    (TileLayerMetadata(cellType, layout, extent, LatLng, bounds), minCol, maxRow)
+  }
+
+  /**
+    * Read particular rasters out of the GeoWave database.  The
+    * particular rasters to read are given by the result running the
+    * provided LayerQuery.
+    *
+    * @param  id               The LayerId specifying the name and zoom level to query
+    * @param  rasterQuery      Produces a list of rasters to read
+    * @param  numPartitions    The number of Spark partitions to use
+    * @param  filterIndexOnly  ?
+    */
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: LayerId, rasterQuery: LayerQuery[K, M], numPartitions: Int, filterIndexOnly: Boolean) = {
+    import GeowaveLayerReader._
+
+    /* Perform checks */
+    if (!attributeStore.layerExists(id))
+      throw new LayerNotFoundError(id)
+    implicitly[ClassTag[K]].toString match {
+      case "geotrellis.spark.SpatialKey" =>
+      case t: String => throw new Exception("Unsupported Key Type: $t")
+    }
+
+    /* Boilerplate */
+    val LayerId(name, zoom) = id
+    val adapter = adapters.filter(_.getCoverageName == name).head
+    val strategy = substrats(zoom)
+    val ranges = strategy.getIndexStrategy.getHighestPrecisionIdRangePerDimension
+    val customIndex = new CustomIdIndex(strategy.getIndexStrategy, index.getIndexModel, index.getId)
+
+    /* GeoTrellis metadata */
+    val (_md, minCol, maxRow) = computeSpatialMetadata(adapter, ranges)
+    val md = _md.asInstanceOf[M]
+
+    /* GeoWave Query and Query Options */
+    val queryOptions = new QueryOptions(adapter, customIndex)
+    val query = {
+      val fn = keyBoundsToGeometry(_md.mapTransform, _: KeyBounds[SpatialKey])
+      val kbs = rasterQuery(md)
+
+      implicitly[ClassTag[K]].toString match {
+        case "geotrellis.spark.SpatialKey" => { // Spatial Query
```

Comment:
Though I think we want to get rid of this completely, but imagine we renamed SpatialKey...it would be much better to match against something from the actual `implicitly[ClassTag[SpatialKey]]`